hbase-backup/src/main/java/org/apache/hadoop/hbase/backup/replication/ContinuousBackupStagedHFileCleaner.java

```diff
@@ -0,0 +1,148 @@
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.backup.replication;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.Abortable;
+import org.apache.hadoop.hbase.HBaseInterfaceAudience;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+import org.apache.hadoop.hbase.master.cleaner.BaseHFileCleanerDelegate;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hbase.thirdparty.com.google.common.collect.Iterables;
+
+/**
+ * A custom HFile cleaner delegate for continuous backup scenarios in HBase. This cleaner prevents
+ * the deletion of HFiles that are staged for bulk loading as part of the continuous backup process.
+ * It interacts with the HBase `StagedBulkloadFileRegistry` to determine which files should be
+ * retained.
+ * <p>
+ * Implements the {@link BaseHFileCleanerDelegate} for integrating with the HBase cleaner framework
+ * and the {@link Abortable} interface to handle error scenarios gracefully.
+ * </p>
+ */
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.CONFIG)
+public class ContinuousBackupStagedHFileCleaner extends BaseHFileCleanerDelegate
+  implements Abortable {
+  private static final Logger LOG =
+    LoggerFactory.getLogger(ContinuousBackupStagedHFileCleaner.class);
+
+  private boolean stopped = false;
+  private boolean aborted;
+  private Configuration conf;
+  private Connection connection;
+
+  @Override
+  public void setConf(Configuration config) {
+    this.conf = config;
+    this.connection = null;
+    try {
+      // Establishing HBase connection
+      this.connection = ConnectionFactory.createConnection(conf);
+      LOG.info("HBase connection established successfully.");
+    } catch (IOException ioe) {
+      LOG.error("Couldn't establish connection to HBase", ioe);
+    }
+  }
+
+  @Override
+  public Iterable<FileStatus> getDeletableFiles(Iterable<FileStatus> files) {
+    if (conf == null) {
+      LOG.warn("Configuration is not set. Returning original list of files.");
+      return files;
+    }
+
+    if (connection == null) {
+      try {
+        connection = ConnectionFactory.createConnection(conf);
+        LOG.info("HBase connection re-established in getDeletableFiles.");
+      } catch (IOException e) {
+        LOG.error("Failed to re-establish HBase connection. Returning no deletable files.", e);
+        return Collections.emptyList();
+      }
+    }
+
+    try {
+      // Fetch staged files from HBase
+      Set<String> stagedFiles = StagedBulkloadFileRegistry.listAllBulkloadFiles(connection);
+      LOG.debug("Fetched {} staged files from HBase.", stagedFiles.size());
```

Comment:
Do we really need to maintain the list of staged files in an HBase table?
Why not just return false (do not delete) if the file is in the staging area?

Reply:
Use special file name prefix while staging:
```
bulkload1.hfile
dontdelete_bulkload2.hfile
dontdelete_bulkload3.hfile
```
`bulkload1.hfile` can be deleted from staging area others should be kept. Rename operation is atomic in HDFS afaik.

Reply:
For the bulkload files, we are not moving them anywhere. They remain in the data directory or archive directory. We are simply maintaining a reference to them in the HBase table.

Reply:
I suggested another way (file renaming) to avoid accidentally deleting them. This way we could avoid maintaining an HBase table.